### PR TITLE
Postgres - common changes

### PR DESCRIPTION
**Pull request by Dharmil-searce**
Refactored convert.go to make functions public which would be useful to make a separate postgres package

### CODE REVIEW

**Comment by Dharmil-searce** [comment]
@googlebot I signed it!

**Comment by hengfengli** [comment]
@Dharmil-searce Is this the first step in our planning doc? It seems that it includes the first two steps in this PR. Or maybe I am wrong.

**Comment by Dharmil-searce** [comment]
@hengfengli Yes, it includes 2 steps from the doc. 1st commit changes are useful for moving Postgres in different package. Only 1st commit standalone is not useful and justified. That is the reason I have added 2 steps in this PR for 'postgres package'. If you suggest then we can also do different PR for each commit.

**Comment by hengfengli** [comment]
I would prefer to split them into separate PRs so that each PR can become smaller and better for code reviews. Thanks for doing that.

**Comment by Dharmil-searce** [comment]
@hengfengli Please check. I have removed 2nd commit and I will make separate PR for postgres package.

File: internal/convert.go [hunk 1]
```diff
@@ -373,38 +419,66 @@ func (conv *Conv) getStatementStat(s string) *statementStat {
...
+func (conv *Conv) SchemaMode() bool {
 	return conv.mode == schemaOnly
 }
 
-func (conv *Conv) dataMode() bool {
+func (conv *Conv) DataMode() bool {
 	return conv.mode == dataOnly
 }
+
+func byteSize(r *row) int64 {
```
Comment (nevinheintze) [inline internal/convert.go:475]: Just noting that byteSize used to be defined in data.go and used in convert.go.
In short, this is a cleanup, and allows us to keep byteSize private.

File: internal/convert.go [hunk 2]
```diff
@@ -201,6 +203,11 @@ func (conv *Conv) Rows() int64 {
 	return n
 }
 
+// GetRows return count of data rows for a srcTable
+func (conv *Conv) GetRows(srcTable string) int64 {
```
Comment (nevinheintze) [inline internal/convert.go:207]: In general, we've avoided adding accessors to convert.go, and instead have directly accessed the subfields of conv as needed. Note that there are functions like BadRows and Statements, but they all do some computation e.g. sum up bad rows etc.

For consistency I think we should continue down that path. That means we'd have to make stats public. Given that much of Conv is being made public anyway, that seems reasonable.

In other words, if anything in convert.go would need an accessor, let's just make it public and avoid the accessor. Same comment for the other GetNNN functions added below.

Hengfeng -- do you have any thoughts on this?

File: internal/convert.go [hunk 3]
```diff
@@ -311,57 +342,72 @@ func (conv *Conv) unexpected(u string) {
...
 	if b {
 		conv.stats.rows[srcTable]++
 	}
 }
 
-func (conv *Conv) statsAddRows(srcTable string, count int64) {
+// StatsAddRows increments the count of rows for 'srcTable' by the count specified
+func (conv *Conv) StatsAddRows(srcTable string, count int64) {
 	conv.stats.rows[srcTable] += count
 }
```
Comment (nevinheintze) [inline internal/convert.go:362]: Suggest dropping this function and just inlining the access to conv.stats.rows (assuming we make stats public) where the function is used.

File: internal/convert.go [hunk 4]
```diff
@@ -311,57 +342,72 @@ func (conv *Conv) unexpected(u string) {
...
 // statsAddGoodRow increments the good-row stats for 'srcTable' if b
-// is true.  See statsAddRow comments for context.
+// is true.  See StatsAddRow comments for context.
 func (conv *Conv) statsAddGoodRow(srcTable string, b bool) {
 	if b {
 		conv.stats.goodRows[srcTable]++
 	}
 }
 
-func (conv *Conv) statsAddGoodRows(srcTable string, count int64) {
```
Comment (nevinheintze) [inline internal/convert.go:340]: Same here.

File: internal/convert.go [hunk 5]
```diff
@@ -373,38 +419,66 @@ func (conv *Conv) getStatementStat(s string) *statementStat {
 	return conv.stats.statement[s]
 }
 
-func (conv *Conv) skipStatement(l []nodes.Node) {
-	if conv.schemaMode() { // Record statement stats on first pass only.
+func (conv *Conv) SkipStatement(l []nodes.Node) {
```
Comment (nevinheintze) [inline internal/convert.go:422]: Yeah, the existing code here is kind of broken: we shouldn't reference nodes.Node in convert.go.
All of that should live in pgdump.go.

I think what should happen longer-term is that prNodes and prNodeType should be moved into pgdump.go.

Let's go with what you have for the moment, but could you add a TODO something like:

// TODO: Move prNodes to pgdump.go, remove all references to nodes.Node in convert.go,
// and just have one version of SkipStatement, ErrorInStatement, SchemaStatement and DataStatement.

File: internal/data.go [hunk 6]
```diff
@@ -380,14 +380,3 @@ func processQuote(s string) (string, error) {
 	}
 	return s, nil
 }
-
-func byteSize(r *row) int64 {
-	n := int64(len(r.table))
```
Comment (nevinheintze) [inline internal/data.go:385]: Just noting: this is now in convert.go.

File: internal/convert.go [hunk 7]
```diff
@@ -373,38 +419,66 @@ func (conv *Conv) getStatementStat(s string) *statementStat {
...
+	if conv.SchemaMode() { // Record statement stats on first pass only.
 		s := prNodes(l)
 		VerbosePrintf("Skipping statement: %s\n", s)
 		conv.getStatementStat(s).skip++
 	}
 }
+func (conv *Conv) SkipStatementMySQL(stmtType string) {
```
Comment (hengfengli) [inline internal/convert.go:429]: Is it possible to merge `SkipStatement` and `SkipStatementMySQL`? What I mean is to move `s := prNodes(l)` out from this function and then `SkipStatement` should work for both `Postgres` and `Mysql`. Same for `ErrorInStatement ` and `ErrorInStatementMYSQL`. 

If this has been done in next PRs, then I am fine with this right now.

File: internal/convert.go [hunk 8]
```diff
@@ -134,6 +135,7 @@ func MakeConv() *Conv {
 			statement:  make(map[string]*statementStat),
 			unexpected: make(map[string]int64),
 		},
+		Offset: "",
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply internal/convert.go:138]: MySQL timezone are just offsets, example "+05:30". The statement to set timezone in mysql is like `SET time_zone = "+9:50";` and we would not be getting a timezone ID as in postgres. Thus, we will capture the value of offset while parsing SET statement and use conv.Offset for timestamp conversion.

File: internal/convert.go [hunk 9]
```diff
@@ -201,6 +203,11 @@ func (conv *Conv) Rows() int64 {
 	return n
 }
 
+// GetRows return count of data rows for a srcTable
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply internal/convert.go]: Removed GetX() function in a new commit

File: internal/convert.go (hunk 5 above)
Comment (Dharmil-searce) [reply internal/convert.go]: I had created new functions with MySQL suffix because using one version needed changes in pgdump.go.
Sure, now you can check in new commit that I have moved prNodes and prNodeType to pgdump.go and created one version for SkipStatement, ErrorInStatement, SchemaStatement and DataStatement functions.

File: internal/convert.go (hunk 3 above)
Comment (Dharmil-searce) [reply internal/convert.go]: Agreed. Now that we are making Stats as public, functions except the one's using bool parameter are not required. I have made changes accordingly.

File: internal/convert.go [hunk 10]
```diff
@@ -224,17 +231,36 @@ func (conv *Conv) Statements() int64 {
...
 	return int64(len(conv.stats.unexpected))
 }
 
+// GetUnexpected returns all count of unexpected conditions,
+// broken down by condition description
+func (conv *Conv) GetUnexpected() map[string]int64 {
+	return conv.stats.unexpected
+}
+
+func (conv *Conv) StatsAddUnexpecteds(u string, count int64) {
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply internal/convert.go]: I removed this function as we have made Unexpected public now.

File: internal/convert.go [hunk 11]
```diff
@@ -263,34 +289,39 @@ func (conv *Conv) SampleBadRows(n int) []string {
...
 	}
 }
 
 // SetLocation configures the timezone for data conversion.
 func (conv *Conv) SetLocation(loc *time.Location) {
-	conv.location = loc
+	conv.Location = loc
+}
+
+// SetOffset configures the offset for data conversion.
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply internal/convert.go]: Ok, I removed this function and instead will use conv.Offset.

File: internal/convert.go (hunk 7 above)
Comment (Dharmil-searce) [reply internal/convert.go]: I didn't merge it at first because it needed changes in pgdump. But now after your suggestion, I have merged these functions in latest commit.

File: internal/convert.go (hunk 2 above)
Comment (Dharmil-searce) [reply internal/convert.go]: Ok, I have made stats and it's fields public and removed accessors i.e GetNNN functions.

File: internal/convert.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hengfengli) [reply internal/convert.go]: Then, we should be more specific to change its name to `TimezoneOffset`.

**Comment by hengfengli** [comment]
@Dharmil-searce Thanks for resolving the comments. It can be merged once @nevinheintze approves it.

**Comment by Dharmil-searce** [comment]
@nevinheintze Sure, we can discuss on our weekly sync up call. Handling of timezone is different in mysql and postgres but sure I will check on how we can have only one timezone spec in convert.go.
@hengfengli I have renamed renaming Offset to TimezoneOffset.
PR can me merged now
